DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp

```diff
@@ -0,0 +1,259 @@
+                     double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
+      h.single_task<ContractFastKernel>([=]() [[intel::kernel_args_restrict]] {
+#pragma clang fp contract(fast)
+        double temp1 = 0.0, temp2 = 0.0;
+        temp1 = accessorA[0] + accessorB[0];
+        temp2 = accessorC[0] + accessorD[0];
+        accessorRes[0] = temp1 * temp2;
+      });
+    });
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+}
+
+// Launch a kernel on the device specified by selector.
+// The kernel's functionality is designed to show the
+// performance impact of the fp reassociate pragma.
+void enable_reassociate(const device_selector &selector, const double a,
+                        const double b, const double c, const double d,
+                        double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
+      h.single_task<ReassociateOnKernel>([=]() [[intel::kernel_args_restrict]] {
+#pragma clang fp reassociate(on)
```

Comment:
Please add comment here explaining what this pragma allows the compiler to do to the floating point operations below.